### DISCUSSION TITLE
Update/Replace/AddAttachment to draft but keep the unique message id

### DISCUSSION THREAD

**Discussion by snailcatcher** [jstedfast/MailKit#1845] [Q&A] [answered]
Hi there,

I want to add an attachment to a existing draft email and sync the change back to the email server without sending the email.

I ended up with this code:
```c#
 var folder = await _imapClient.GetFolderAsync(folderPath, cancellationToken);
await folder.OpenAsync(FolderAccess.ReadWrite, cancellationToken);

var message = await folder.GetMessageAsync(uniqueId.Value, cancellationToken);

if (message == null)
   throw new EmailNotFoundException($"The draft email with id '{emailId}' could not be found.");

var builder = new BodyBuilder {
   TextBody = message.TextBody,
   HtmlBody = message.HtmlBody
};
await builder.Attachments.AddAsync(attachmentFileName, attachmentStream, cancellationToken);
message.Body = builder.ToMessageBody();

// and here is the problem
var messageId = await folder.ReplaceAsync(uniqueId.Value, message, MessageFlags.Draft, cancellationToken);
if (!messageId.HasValue)
   throw new EmailCreationException("The email draft could not be created.");

Debug.Assert(uniqueId.Value.Equals(messageId.Value), "The unique id of the message should not change.");
// this leads to an update of the draft, but also changes the unique id, which I want to keep
```

Is there another/better way to update the draft without changing the uniqueId of my email?

The reason I want to archive this, is because i want to add multiple attachments to one email before sending it. Each of this attachments is provided one by one. So I want to use the mail server as kind of storage for the "complete" email draft I build with all the information and attachments. And once the mail draft is complete I will send it via smtp. 

Sincerly snailcatcher

**Comment by jstedfast** [comment]
The way you are doing it is the recommended way.

The UID changing is a side effect, most likely, of the server not supporting the REPLACE command, so MailKit has to work-around that by appending a *new* message to the Drafts folder (which means it'll get a new UID).

Clients cannot override the server's choice of UID.

**Reply by snailcatcher** [reply]
So this means it is impossible to keep the unique id and I have to use the new one after executing the *Replace* function?
This is not what I hoped for but at least its an answer. Thank you!

**Reply by snailcatcher** [reply]
But now I have another problem. If I use my code above multiple times I only get a draft with the last attachment I added.
So is it really necessary to download each of the attachments of the draft and append the new attachment to it just to upload everything again.
This seems very inefficient, is there no better why to solve this?

**Reply by jstedfast** [reply, marked as answer]
If you want to incrementally add attachments, it would be best to download the entire message and add them w/o the use of the BodyBuilder (BodyBuilder, honestly, is for constructing *new* messages and not *adding* to existing messages).

The BodyBuilder.ToMessageBody() method *replaces* the existing message body with the new content.

You can do something more like this:

```csharp
var message = await folder.GetMessageAsync(uniqueId.Value, cancellationToken);

if (message == null)
   throw new EmailNotFoundException($"The draft email with id '{emailId}' could not be found.");

// Note: we want to add attachments but make no other change to the message. Since attachments
// are typically added to the top-level multipart/mixed, let's find that multipart -or- add a multipart/mixed
// wrapper around the existing message body if it doesn't already exist.
var multipart = message.Body as Multipart;
Multipart mixed = null;

if (multipart != null && multipart.ContentType.IsMimeType ("multipart", "mixed")) {
    mixed = multipart;
} else {
    mixed = new Multipart ("mixed");
    mixed.Add (message.Body);
    message.Body = mixed;
}

// Now add the attachment(s)...
var mimeType = MimeTypes.GetMimeType (attachmentFileName);

// Note we copy the attachmentStream here because I'm not sure if you are disposing it before you upload the message,
// but you might be able to get away without doing this...
var contentStream = new MemoryStream ();
await attachmentStream.CopyToAsync (contentStream, 4096);
contentStream.Position = 0;

var attachment = new MimePart (mimeType) {
    FileName = Path.GetFileName (attachment.FileName),
    ContentTransferEncoding = ContentEncoding.Base64,
    Content = new MimeContent (contentStream)
};

mixed.Add (attachment);

// replace the draft message
var messageId = await folder.ReplaceAsync(uniqueId.Value, message, MessageFlags.Draft, cancellationToken);
```

**Reply by snailcatcher** [reply]
Just for completeness of this discussion for someone who wants t stay with the BodyBuilder. The following also works 
```c#
...
var builder = new BodyBuilder {
   TextBody = message.TextBody,
   HtmlBody = message.HtmlBody
};

foreach (var attachment in message.Attachments)
{
   builder.Attachments.Add(attachment);
}
await builder.Attachments.AddAsync(attachmentFileName, attachmentStream, cancellationToken);
...
```
Thank you @jstedfast
